### ISSUE TITLE
Unexpected const assertion behavior when used directly inside ternary expression

### ISSUE THREAD

**Issue by btoo** [microsoft/TypeScript#41143]
**TypeScript Version:**  v4.1.0-dev.20201015

**Search Terms:** type inference const assertion ternary expression object literal normalization union

## Case 1
**Code**
```typescript
const emptyObject = {} as const;

const unexpected = (bool: boolean) => bool
  ? emptyObject
  : { maybe: true } as const

const unexpectedResult = unexpected(true)
if ('maybe' in unexpectedResult) {
  unexpectedResult.maybe;
}
```
**Expected behavior:**
- `unexpected`'s return type is inferred as `{} | { readonly maybe: true }`
- `unexpectedResult`'s type is narrowed from `{} | { readonly maybe: true }` to `{ readonly maybe: true }` in `if ('maybe' in unexpectedResult)`

**Actual behavior:**
- `unexpected`'s return type is inferred as _only_ `{}`
- `unexpectedResult.maybe` throws a compilation error

[**Playground Link**](https://www.typescriptlang.org/play?ts=4.1.0-dev.20201015#code/MYewdgzgLgBApgWwA5QJ4HkBGArOxYC8MA3gL4wCGEMokUA3AFCO3QwCuYcAHknlHAAmMIgApMIEABsAXDAnS4FMAEoRAPnmSpjGDAD8JGAgqpMcOVABO7ODHJUa4aLphzEKDDn7NWsTjx8+EIASnAQ7FKEHFy8-EKi1rYqjACWAGYwogDkJmZw2TCpYDGB8YJhEVFqxK4BccEV4ZFQAHR55kykQA)

## Case 2
**Code**
```typescript
const unexpected = (bool: boolean) => bool
  ? { always: true, maybe: true } as const
  : { always: true } as const;

const unexpectedResult = unexpected(true);

if ('maybe' in unexpectedResult) {
  unexpectedResult.maybe
}
```
**Expected behavior:** `typeof unexpectedResult.maybe` inferred as `true`

**Actual behavior:** `typeof unexpectedResult.maybe` inferred as `true | undefined`

[**Playground Link**](https://www.typescriptlang.org/play?ts=4.1.0-dev.20201015#code/MYewdgzgLgBArmApgDwA6OFRATGBeGACgCMQQAbALhlIsQEMwBKfAPhrPICgYYB+GAG8Y9cgHd6ATwjUoAJziIANDAC2U4olkLEMAL4iIMUJCg8Y1YaInTti-YePhoAbi5cT0eEjQYs2ACVECDhyWAIEFHRMHEJ5RSY3LgBLADMiAHJ1SU0MmGSwbyi-HCCQsJZBc0jfGMDg0KgAOmzNLj0uIA) - Interestingly, this code will begin to behave as expected if you simply [move _at least one_ of the inlined `const` assertions into its own variable declarations](https://www.typescriptlang.org/play?ts=4.1.0-dev.20201015&ssl=4&ssc=5&pln=4&pc=43#code/MYewdgzgLgBApgDynATmAhgGwMLmgQQglSgEtwYBeGAbxiwHd0BPCALhihQFc4YBfehBihIUAFDjR0eAgAOcYMgAmVGAAoARiBCYO23XHRgAlFQB8MA5nEwYAflr1MTVhy68ANDAC2LTXDuPHyC6MLSEnYciMhoWLhihMQoZOAA3JIRsgpKcMoASnAQ3Jiw1Ig5KuoecCaSpABmGgDkfswBzTCkYNmKKoXFpWY0tr25BUUlUAB0bQHi-OJAA)

---

**Related Issues:** https://github.com/microsoft/TypeScript/pull/19513

**Comment by MartinJohns** [comment]
Please respect the issue templates for [bug reports](https://raw.githubusercontent.com/microsoft/TypeScript/master/.github/ISSUE_TEMPLATE/Bug_report.md) and [feature requests](https://raw.githubusercontent.com/microsoft/TypeScript/master/.github/ISSUE_TEMPLATE/Feature_request.md). This will make the work for the TypeScript team much easier. :-)

---

The behavior you see is intentional. See #19513 for details.

**Comment by btoo** [comment]
@MartinJohns thanks for the callout - just updated the description to better comply with template

---

> The behavior you see is intentional

Funnily enough, applying [the rules](https://github.com/microsoft/TypeScript/pull/19513) to Case 1 should've yielded `unexpected`'s return type to be `{ readonly maybe?: true | undefined }` but that didn't happen in the playground link. Is this a separate issue?

Assuming for a second that Case 1's `unexpected` _does_ return `{ readonly maybe?: true | undefined }`, then I wonder if this  and Case 2 would be the basis for an argument against object normalization and/or in support of a feature request for preserving constituents' type information when inferring a union. For me, narrowing unions has been one of TypeScript's most powerful ([and oldest](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html)) features, and object normalization prevents me from using it!

**Comment by RyanCavanaugh** [comment]
There's probably confusion here stemming from the fact that `{}` does not *guarantee* that an object has zero properties (object types *are not sealed*). It'd be perfectly legal to write this code, for example:
```ts
const emptyObject: {} = { maybe: "no" };

const unexpected = (bool: boolean) => bool
  ? { maybe: 42 }
  : emptyObject

const unexpectedResult = unexpected(true)
if ('maybe' in unexpectedResult) {
  unexpectedResult.maybe.toFixed();
}
```
which from TypeScript's perspective is *exactly the same* as yours, but this code crashes on the line underlined as an error (which is a pretty good sign that things are going according to plan).

Optional property normalization can *only* correctly occur on fresh object literals; any aliasing in the interim creates the possibility that the property you're testing for with `in` actually does exist and has some other type than the one you expect

**Comment by btoo** [comment]
Thank you for the answer @RyanCavanaugh! Using your insight, I found a way to prevent unwanted object normalizations by pseudo-sealing objects using `{ readonly [key in any]: never } & as_const_type`

[Fixed Case 1](https://www.typescriptlang.org/play?ts=4.1.0-dev.20201015#code/MYewdgzgLgBApgWwA5QJ4HkBGArOxYC8MA3gL4wCGEJMATnBQCbgA2qMA2gNZzsCWYSmFQBdAFwwwcAG5xaMUgG4AUMtCRYAVykAPJHihxGMIgApMIECwkWrDMAEoTAPhi2WymDAD8NBBVRMOAkoWk04BUpqdWhPGAlEFAwcA1UYrV19fCMAJTgITRZCGG04PQMjU1Dwh2U+ADMYUwByf0C4ZpgBEsyKxjyCoqdiONLy7P78wqgAOjaglVIgA)

[Fixed Case 2](https://www.typescriptlang.org/play?ts=4.1.0-dev.20201015#code/MYewdgzgLgBArmApgDwA6OFRATGBeGACgCMQQAbALhlIsQEMwBKfAPhrPICgYYB+GAG8Y9cgHd6ATwjUoAJziIANDAC2U4olkLEMAL4iIMUJCg8Y1YaInTti-YaEw5DbOHKSYAbQDWiTwCWYCJgkgC61EgAbohyDgBkTi70bmAeIuJSMjDy9noA3FxcJtDwSGgYWNgASogQcOSwBAgo6Jg4hLmITIVcAQBmRADk6pKaQzBBZa2VOLX1jSyC5i0V7TV1DVAAdKOaXHpcQA)

However, this solution comes with the disadvantage of needing to manually **_A_**) re-construct the type definition and **_B_**) prepend `{ readonly [key in any]: never } &` every time. **_A_** would have otherwise been handled by `as const`, and **_B_** is what seals the object.

Is it conceivable that, in a future version of TypeScript, `as const` assertions simulate `Object.seal` by including `{ readonly [key in any]: never } &` with the produced type? A feature like this would've prevented my "issue", and it also feels sensible for `as const` types to be sealed. `{ readonly [key in any]: never }` seems to do a good job of this on the type level (more testing should probably be done anyways 😅)

**Comment by MartinJohns** [comment]
@btoo It sounds to me that you want "Exact Types": #12936
